howdoi/plugins/base.py

```diff
@@ -0,0 +1,166 @@
+else:
+    from urllib.request import getproxies
+    from urllib.parse import quote as url_quote, urlparse, parse_qs
+
+    def u(x):
+        return x
+
+# rudimentary standardized 3-level log output
+
+
+def _print_err(x):
+    print("[ERROR] " + x)
+
+
+_print_ok = print  # noqa: E305
+
+
+def _print_dbg(x):
+    print("[DEBUG] " + x)  # noqa: E302
+
+
+CACHE_EMPTY_VAL = "NULL"
+CACHE_DIR = appdirs.user_cache_dir('howdoi')
+CACHE_ENTRY_MAX = 128
+
+if os.getenv('HOWDOI_DISABLE_CACHE'):
+    cache = NullCache()  # works like an always empty cache
+else:
+    cache = FileSystemCache(CACHE_DIR, CACHE_ENTRY_MAX, default_timeout=0)
+
+ANSWER_HEADER = u('{2}  Answer from {0} {2}\n{1}')
+STAR_HEADER = u('\u2605')
+CACHE_EMPTY_VAL = "NULL"
+NO_ANSWER_MSG = '< no answer given >'
+
+if os.getenv('HOWDOI_DISABLE_SSL'):  # Set http instead of https
+    SCHEME = 'http://'
+    VERIFY_SSL_CERTIFICATE = False
+else:
+    SCHEME = 'https://'
+    VERIFY_SSL_CERTIFICATE = True
+
+BLOCK_INDICATORS = (
+    'form id="captcha-form"',
+    'This page appears when Google automatically detects requests coming from your computer '
+    'network which appear to be in violation of the <a href="//www.google.com/policies/terms/">Terms of Service'
+)
+
+URL = os.getenv('HOWDOI_URL') or 'stackoverflow.com'
+
+SEARCH_URLS = {
+    'bing': SCHEME + 'www.bing.com/search?q=site:{0}%20{1}&hl=en',
+    'google': SCHEME + 'www.google.com/search?q=site:{0}%20{1}&hl=en',
+    'duckduckgo': SCHEME + 'duckduckgo.com/?q=site:{0}%20{1}&t=hj&ia=web'
+}
+
+
+class BasePlugin():
+    def __init__(self, cache=None):
+        if cache is None:
+            cache = NullCache()
+        self.cache = cache
+
+    def _is_blocked(self, page):
+        for indicator in BLOCK_INDICATORS:
+            if page.find(indicator) != -1:
+                return True
+        return False
+
+    def _add_links_to_text(self, element):
+        hyperlinks = element.find('a')
+
+        for hyperlink in hyperlinks:
+            pquery_object = pq(hyperlink)
+            href = hyperlink.attrib['href']
+            copy = pquery_object.text()
+            if (copy == href):
+                replacement = copy
+            else:
+                replacement = "[{0}]({1})".format(copy, href)
+            pquery_object.replace_with(replacement)
+
+    def get_link_at_pos(self, links, position):
+        if not links:
+            return False
+        if len(links) >= position:
+            link = links[position - 1]
+        else:
+            link = links[-1]
+        return link
+
+    def get_text(self, element):
+        ''' return inner text in pyquery element '''
+        self._add_links_to_text(element)
+        try:
+            return element.text(squash_space=False)
+        except TypeError:
+            return element.text()
+
+    def _get_search_url(self, search_engine):
+        return SEARCH_URLS.get(search_engine, SEARCH_URLS['google'])
+
+    def _extract_links_from_bing(self, html):
```

Comment:
This can also be a static method.
```suggestion
    @staticmethod
    def _extract_links_from_bing(html):
```